rust/ql/lib/codeql/rust/elements/internal/PathResolution.qll

```diff
@@ -0,0 +1,464 @@
+ * `fileName.rs`, inside the folder `parent`.
+ */
+private predicate modImport(Module m, string fileName, string name, Folder parent) {
+  exists(File f |
+    f = m.getFile() and
+    not m.hasItemList() and
+    // TODO: handle
+    // ```
+    // #[path = "foo.rs"]
+    // mod bar;
+    // ```
+    not m.getAnAttr().getMeta().getPath().getPart().getNameRef().getText() = "path" and
+    name = m.getName().getText() and
+    parent = f.getParentContainer() and
+    fileName = f.getStem()
+  )
+}
+
+/** Holds if `m` is a `mod name;` item importing file `f`. */
+private predicate fileImport(Module m, SourceFile f) {
+  exists(string fileName, string name, Folder parent | modImport(m, fileName, name, parent) |
+    // sibling import
+    fileModule(f, name, parent)
+    or
+    // child import
+    fileModule(f, name, parent.getFolder(fileName))
+  )
+}
+
+/**
+ * Holds if `mod` is a `mod name;` item targeting a file resulting in `item` being
+ * in scope under the name `name`.
+ */
+private predicate fileImportEdge(Module mod, string name, ItemNode item) {
+  item.isPublic() and
+  exists(SourceFile f |
+    fileImport(mod, f) and
+    sourceFileEdge(f, name, item)
+  )
+}
+
+pragma[nomagic]
+private predicate useTreeIsGlobImport(UseTree use) {
+  // TODO: the extractor should provide this information
+  use.getLocation() != use.getPath().getLocation() and
+  not use.hasUseTreeList() and
+  not use.hasRename()
+}
+
+private predicate useTreeDeclares(UseTree tree, string name) {
+  not useTreeIsGlobImport(tree) and
+  not exists(tree.getUseTreeList()) and
+  (
+    name = tree.getRename().getName().getText() and
+    name != "_"
+    or
+    not tree.hasRename() and
+    name = tree.getPath().getPart().getNameRef().getText()
+  )
+  or
+  exists(UseTree mid |
+    useTreeDeclares(mid, name) and
+    mid = tree.getUseTreeList().getAUseTree()
+  )
+}
+
+/**
+ * Holds if `item` explicitly declares a sub item named `name`. This includes
+ * items declared by `use` statements, except for glob imports.
+ */
+pragma[nomagic]
+private predicate declares(ItemNode item, string name) {
+  exists(ItemNode child | child.getImmediateParent() = item |
+    child.getName() = name
+    or
+    useTreeDeclares(child.(Use).getUseTree(), name)
+  )
+  or
+  exists(MacroCallItemNode call |
+    declares(call, name) and
+    call.getImmediateParent() = item
+  )
+}
+
+private class RelevantPath extends Path {
+  RelevantPath() { not this = any(VariableAccess va).(PathExpr).getPath() }
+
+  pragma[nomagic]
+  predicate isRoot(string name) {
+    not exists(this.getQualifier()) and
+    not this = any(UseTreeList list).getAUseTree().getPath() and
+    name = this.getPart().getNameRef().getText()
+  }
+}
+
+/**
+ * Holds if the root path `root` references an item named `name`, and `name`
+ * may be looked up inside enclosing item `encl`.
+ */
+pragma[nomagic]
+private predicate rootPathLookup(RelevantPath root, string name, ItemNode encl) {
```

Comment:
A lot of the other predicates here have `ItemNode` as a result, could this one be the same?

Reply:
I think I prefer it as-is, because `encl` is not functionally determined.